File: sora-2-api-asynchronous/image-or-text-to-video/sora-2-async.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"mime/multipart"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"time"
)

// 配置参数
const (
	BaseURL      = "https://api.apiyi.com/v1/videos"
	APIKey       = "sk-"
	Prompt       = "一只橘色的小猫在阳光明媚的花园里追逐蝴蝶"
	Model        = "sora-2"
	Size         = "1280x720"
	Seconds      = "15"
	PollInterval = 30 * time.Second // 每 30 秒查询一次
	MaxWaitTime  = 10 * time.Minute // 最长等待 10 分钟
)

// ==================== 模式配置 ====================
// 设置 ImagePath 来切换生成模式：
// - ImagePath = "" (留空)        → 文字生成视频
// - ImagePath = "/path/to/image" → 图片生成视频
// =================================================

const ImagePath = "" // 文字生成视频模式
// const ImagePath = "/path/to/your/image.png" // 图片生成视频模式

// VideoResponse 视频响应结构
type VideoResponse struct {
	ID          string `json:"id"`
	Status      string `json:"status"`
	Progress    int    `json:"progress"`
	URL         string `json:"url"`
	CreatedAt   int64  `json:"created_at"`
	CompletedAt int64  `json:"completed_at"`
}

// 根据文件扩展名推断 MIME 类型
func getMimeType(filePath string) string {
	ext := strings.ToLower(filepath.Ext(filePath))
	mimeTypes := map[string]string{
		".png":  "image/png",
		".jpg":  "image/jpeg",
		".jpeg": "image/jpeg",
	}
	if mime, ok := mimeTypes[ext]; ok {
		return mime
	}
	return "image/png"
}

// 第一步：提交视频生成请求（自动识别文字或图片模式）
func step1SubmitRequest() (string, bool, error) {
	fmt.Println("============================================================")
	fmt.Println("第一步：提交视频生成请求")
	fmt.Println("============================================================")

	// 检查是否为图片生成模式
	isImageMode := ImagePath != "" && fileExists(ImagePath)

	if ImagePath != "" && !isImageMode {
		fmt.Printf("⚠️  警告：图片路径已设置但文件不存在: %s\n", ImagePath)
		fmt.Println("    将使用文字生成视频模式")
	}

	// 创建 multipart form
	body := &bytes.Buffer{}
	writer := multipart.NewWriter(body)

	// 添加基础表单字段
	writer.WriteField("prompt", Prompt)
	writer.WriteField("model", Model)
	writer.WriteField("size", Size)
	writer.WriteField("seconds", Seconds)

	if isImageMode {
		// 图片生成视频模式
		fmt.Println("📷 模式：图片生成视频")
		fmt.Println("📤 发送请求...")
		fmt.Printf("   - 提示词: %s\n", Prompt)
		fmt.Printf("   - 图片: %s\n", ImagePath)
		fmt.Printf("   - 模型: %s\n", Model)
		fmt.Printf("   - 尺寸: %s\n", Size)
		fmt.Printf("   - 时长: %s秒\n", Seconds)

		// 添加图片文件
		file, err := os.Open(ImagePath)
		if err != nil {
			return "", false, fmt.Errorf("打开文件失败: %v", err)
		}
		defer file.Close()

		part, err := writer.CreateFormFile("input_reference", filepath.Base(ImagePath))
		if err != nil {
			return "", false, fmt.Errorf("创建表单文件失败: %v", err)
		}
		io.Copy(part, file)
	} else {
		// 文字生成视频模式
		fmt.Println("📝 模式：文字生成视频")
		fmt.Println("📤 发送请求...")
		fmt.Printf("   - 提示词: %s\n", Prompt)
		fmt.Printf("   - 模型: %s\n", Model)
		fmt.Printf("   - 尺寸: %s\n", Size)
		fmt.Printf("   - 时长: %s秒\n", Seconds)
	}

	writer.Close()

	// 发送请求
	req, err := http.NewRequest("POST", BaseURL, body)
	if err != nil {
		return "", false, err
	}

	req.Header.Set("Authorization", APIKey)
	req.Header.Set("Content-Type", writer.FormDataContentType())

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return "", false, err
	}
	defer resp.Body.Close()

	respBody, _ := io.ReadAll(resp.Body)

	if resp.StatusCode == 200 {
		var result VideoResponse
		if err := json.Unmarshal(respBody, &result); err != nil {
			return "", false, err
		}

		fmt.Println("✅ 请求提交成功!")
		fmt.Printf("   - 视频ID: %s\n", result.ID)
		fmt.Printf("   - 状态: %s\n", result.Status)
		fmt.Printf("   - 创建时间: %s\n", time.Unix(result.CreatedAt, 0).Format("2006-01-02 15:04:05"))
		fmt.Println("   ⏱️  预计生成时间: 3-5 分钟")

		return result.ID, isImageMode, nil
	}

	return "", false, fmt.Errorf("❌ 请求失败: %d\n响应: %s", resp.StatusCode, string(respBody))
}

// 第二步：轮询查询视频状态
func step2PollStatus(videoID string) error {
	fmt.Println("\n============================================================")
	fmt.Println("第二步：查询视频生成状态")
	fmt.Println("============================================================")
	fmt.Printf("💡 提示：视频生成通常需要 3-5 分钟，每 %v 秒自动查询一次\n", int(PollInterval.Seconds()))

	startTime := time.Now()
	client := &http.Client{}

	for {
		elapsed := time.Since(startTime)

		// 检查超时
		if elapsed > MaxWaitTime {
			return fmt.Errorf("⏱️ 超时：已等待 %v，停止查询", MaxWaitTime)
		}

		fmt.Printf("\n🔍 查询状态... (已等待 %d 秒)\n", int(elapsed.Seconds()))

		// 查询状态
		statusURL := fmt.Sprintf("%s/%s", BaseURL, videoID)
		req, _ := http.NewRequest("GET", statusURL, nil)
		req.Header.Set("Authorization", APIKey)

		resp, err := client.Do(req)
		if err != nil {
			fmt.Printf("❌ 请求失败: %v\n", err)
			time.Sleep(PollInterval)
			continue
		}

		respBody, _ := io.ReadAll(resp.Body)
		resp.Body.Close()

		if resp.StatusCode == 200 {
			var result VideoResponse
			if err := json.Unmarshal(respBody, &result); err != nil {
				fmt.Printf("❌ 解析响应失败: %v\n", err)
				time.Sleep(PollInterval)
				continue
			}

			fmt.Printf("   - 状态: %s\n", result.Status)
			fmt.Printf("   - 进度: %d%%\n", result.Progress)

			switch result.Status {
			case "completed":
				fmt.Println("✅ 视频生成完成!")
				fmt.Printf("   - 视频URL: %s\n", result.URL)
				fmt.Printf("   - 完成时间: %s\n", time.Unix(result.CompletedAt, 0).Format("2006-01-02 15:04:05"))
				return nil
			case "failed":
				return fmt.Errorf("❌ 视频生成失败")
			default:
				fmt.Printf("⏳ 视频生成中，等待 %d 秒后继续查询...\n", int(PollInterval.Seconds()))
				time.Sleep(PollInterval)
			}
		} else {
			fmt.Printf("❌ 状态查询失败: %d\n", resp.StatusCode)
			time.Sleep(PollInterval)
		}
	}
}

// 第三步：下载视频
func step3DownloadVideo(videoID string, isImageMode bool) error {
	fmt.Println("\n============================================================")
	fmt.Println("第三步：下载视频")
	fmt.Println("============================================================")

	// 根据模式生成文件名
	modePrefix := "text"
	if isImageMode {
		modePrefix = "image"
	}
	outputFile := fmt.Sprintf("sora_%s_video_%s.mp4", modePrefix, videoID[len(videoID)-8:])

	fmt.Println("📥 开始下载视频...")

	contentURL := fmt.Sprintf("%s/%s/content", BaseURL, videoID)
	req, _ := http.NewRequest("GET", contentURL, nil)
	req.Header.Set("Authorization", APIKey)

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return fmt.Errorf("❌ 下载请求失败: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("❌ 下载失败: %d\n响应: %s", resp.StatusCode, string(body))
	}

	// 保存文件
	out, err := os.Create(outputFile)
	if err != nil {
		return fmt.Errorf("❌ 创建文件失败: %v", err)
	}
	defer out.Close()

	size, err := io.Copy(out, resp.Body)
	if err != nil {
		return fmt.Errorf("❌ 保存文件失败: %v", err)
	}

	absPath, _ := filepath.Abs(outputFile)
	fmt.Println("✅ 视频下载成功!")
	fmt.Printf("   - 保存路径: %s\n", absPath)
	fmt.Printf("   - 文件大小: %d 字节\n", size)

	return nil
}

// 辅助函数：检查文件是否存在
func fileExists(path string) bool {
	_, err := os.Stat(path)
	return !os.IsNotExist(err)
}

func main() {
	fmt.Println("\n🎬 开始 Sora-2 视频生成完整流程")
	fmt.Printf("⏰ 开始时间: %s\n\n", time.Now().Format("2006-01-02 15:04:05"))

	// 第一步：提交请求
	videoID, isImageMode, err := step1SubmitRequest()
	if err != nil {
		fmt.Printf("\n❌ 流程终止：%v\n", err)
		return
	}

	// 第二步：轮询状态
	if err := step2PollStatus(videoID); err != nil {
		fmt.Printf("\n❌ 流程终止：%v\n", err)
		return
	}

	// 第三步：下载视频
	if err := step3DownloadVideo(videoID, isImageMode); err != nil {
		fmt.Printf("\n❌ 流程终止：%v\n", err)
		return
	}

	fmt.Println("\n============================================================")
	fmt.Println("🎉 完整流程执行成功!")
	fmt.Println("============================================================")
	fmt.Printf("⏰ 结束时间: %s\n", time.Now().Format("2006-01-02 15:04:05"))
}

File: sora-2-api-asynchronous/image-or-text-to-video/Sora2Async.java
import java.io.*;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;
import org.json.JSONObject;

/**
 * Sora-2 视频生成完整异步流程 - Java 版本
 * 智能双模式：自动识别文字生成视频 / 图片生成视频
 *
 * 依赖: org.json (添加到 pom.xml 或下载 json-20240303.jar)
 * Maven:
 * <dependency>
 *     <groupId>org.json</groupId>
 *     <artifactId>json</artifactId>
 *     <version>20240303</version>
 * </dependency>
 *
 * 编译: javac -cp json-20240303.jar Sora2Async.java
 * 运行: java -cp .:json-20240303.jar Sora2Async
 */
public class Sora2Async {

    // 配置参数
    private static final String BASE_URL = "https://api.apiyi.com/v1/videos";
    private static final String API_KEY = "sk-";
    private static final String PROMPT = "一只橘色的小猫在阳光明媚的花园里追逐蝴蝶";
    private static final String MODEL = "sora-2";
    private static final String SIZE = "1280x720";
    private static final String SECONDS = "15";

    // ==================== 模式配置 ====================
    // 设置 IMAGE_PATH 来切换生成模式：
    // - IMAGE_PATH = null (留空)      → 文字生成视频
    // - IMAGE_PATH = "/path/to/image" → 图片生成视频
    // =================================================

    private static final String IMAGE_PATH = null;  // 文字生成视频模式
    // private static final String IMAGE_PATH = "/path/to/your/image.png";  // 图片生成视频模式

    // 轮询配置
    private static final int POLL_INTERVAL = 30000; // 30 秒
    private static final int MAX_WAIT_TIME = 600000; // 10 分钟

    private static final String BOUNDARY = "----WebKitFormBoundary" + System.currentTimeMillis();

    /**
     * 第一步：提交视频生成请求（自动识别文字或图片模式）
     */
    private static VideoSubmitResult step1SubmitRequest() throws Exception {
        System.out.println("============================================================");
        System.out.println("第一步：提交视频生成请求");
        System.out.println("============================================================");

        // 检查是否为图片生成模式
        boolean isImageMode = IMAGE_PATH != null && new File(IMAGE_PATH).exists();

        if (IMAGE_PATH != null && !isImageMode) {
            System.out.println("⚠️  警告：图片路径已设置但文件不存在: " + IMAGE_PATH);
            System.out.println("    将使用文字生成视频模式");
        }

        URL url = new URL(BASE_URL);
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("POST");
        conn.setDoOutput(true);
        conn.setRequestProperty("Authorization", API_KEY);
        conn.setRequestProperty("Content-Type", "multipart/form-data; boundary=" + BOUNDARY);

        try (OutputStream out = conn.getOutputStream();
             PrintWriter writer = new PrintWriter(new OutputStreamWriter(out, "UTF-8"), true)) {

            // 添加基础表单字段
            addFormField(writer, "prompt", PROMPT);
            addFormField(writer, "model", MODEL);
            addFormField(writer, "size", SIZE);
            addFormField(writer, "seconds", SECONDS);

            if (isImageMode) {
                // 图片生成视频模式
                System.out.println("📷 模式：图片生成视频");
                System.out.println("📤 发送请求...");
                System.out.println("   - 提示词: " + PROMPT);
                System.out.println("   - 图片: " + IMAGE_PATH);
                System.out.println("   - 模型: " + MODEL);
                System.out.println("   - 尺寸: " + SIZE);
                System.out.println("   - 时长: " + SECONDS + "秒");

                // 添加文件
                addFilePart(writer, out, "input_reference", new File(IMAGE_PATH));
            } else {
                // 文字生成视频模式
                System.out.println("📝 模式：文字生成视频");
                System.out.println("📤 发送请求...");
                System.out.println("   - 提示词: " + PROMPT);
                System.out.println("   - 模型: " + MODEL);
                System.out.println("   - 尺寸: " + SIZE);
                System.out.println("   - 时长: " + SECONDS + "秒");
            }

            // 结束边界
            writer.append("--").append(BOUNDARY).append("--").append("\r\n");
        }

        int responseCode = conn.getResponseCode();
        String responseBody = readResponse(conn);

        if (responseCode == 200) {
            JSONObject result = new JSONObject(responseBody);
            String videoId = result.getString("id");
            String status = result.getString("status");
            long createdAt = result.getLong("created_at");

            System.out.println("✅ 请求提交成功!");
            System.out.println("   - 视频ID: " + videoId);
            System.out.println("   - 状态: " + status);
            System.out.println("   - 创建时间: " + formatTimestamp(createdAt));
            System.out.println("   ⏱️  预计生成时间: 3-5 分钟");

            return new VideoSubmitResult(videoId, isImageMode);
        } else {
            throw new IOException("❌ 请求失败: " + responseCode + "\n响应: " + responseBody);
        }
    }

    /**
     * 第二步：轮询查询视频状态
     */
    private static void step2PollStatus(String videoId) throws Exception {
        System.out.println("\n============================================================");
        System.out.println("第二步：查询视频生成状态");
        System.out.println("============================================================");
        System.out.println("💡 提示：视频生成通常需要 3-5 分钟，每 30 秒自动查询一次");

        long startTime = System.currentTimeMillis();

        while (true) {
            long elapsed = System.currentTimeMillis() - startTime;

            // 检查超时
            if (elapsed > MAX_WAIT_TIME) {
                throw new Exception("⏱️ 超时：已等待 " + (MAX_WAIT_TIME / 1000) + " 秒，停止查询");
            }

            System.out.println("\n🔍 查询状态... (已等待 " + (elapsed / 1000) + " 秒)");

            URL url = new URL(BASE_URL + "/" + videoId);
            HttpURLConnection conn = (HttpURLConnection) url.openConnection();
            conn.setRequestMethod("GET");
            conn.setRequestProperty("Authorization", API_KEY);

            int responseCode = conn.getResponseCode();
            String responseBody = readResponse(conn);

            if (responseCode == 200) {
                JSONObject result = new JSONObject(responseBody);
                String status = result.getString("status");
                int progress = result.optInt("progress", 0);

                System.out.println("   - 状态: " + status);
                System.out.println("   - 进度: " + progress + "%");

                switch (status) {
                    case "completed":
                        System.out.println("✅ 视频生成完成!");
                        System.out.println("   - 视频URL: " + result.optString("url"));
                        System.out.println("   - 完成时间: " + formatTimestamp(result.optLong("completed_at")));
                        return;
                    case "failed":
                        throw new Exception("❌ 视频生成失败");
                    default:
                        System.out.println("⏳ 视频生成中，等待 " + (POLL_INTERVAL / 1000) + " 秒后继续查询...");
                        Thread.sleep(POLL_INTERVAL);
                }
            } else {
                System.out.println("❌ 状态查询失败: " + responseCode);
                Thread.sleep(POLL_INTERVAL);
            }
        }
    }

    /**
     * 第三步：下载视频
     */
    private static void step3DownloadVideo(String videoId, boolean isImageMode) throws Exception {
        System.out.println("\n============================================================");
        System.out.println("第三步：下载视频");
        System.out.println("============================================================");

        // 根据模式生成文件名
        String modePrefix = isImageMode ? "image" : "text";
        String outputFile = "sora_" + modePrefix + "_video_" +
                           videoId.substring(Math.max(0, videoId.length() - 8)) + ".mp4";

        System.out.println("📥 开始下载视频...");

        URL url = new URL(BASE_URL + "/" + videoId + "/content");
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("GET");
        conn.setRequestProperty("Authorization", API_KEY);

        int responseCode = conn.getResponseCode();

        if (responseCode == 200) {
            try (InputStream in = conn.getInputStream();
                 FileOutputStream out = new FileOutputStream(outputFile)) {

                byte[] buffer = new byte[8192];
                int bytesRead;
                long totalBytes = 0;

                while ((bytesRead = in.read(buffer)) != -1) {
                    out.write(buffer, 0, bytesRead);
                    totalBytes += bytesRead;
                }

                System.out.println("✅ 视频下载成功!");
                System.out.println("   - 保存路径: " + new File(outputFile).getAbsolutePath());
                System.out.println("   - 文件大小: " + totalBytes + " 字节");
            }
        } else {
            throw new IOException("❌ 下载失败: " + responseCode);
        }
    }

    /**
     * 主流程
     */
    public static void main(String[] args) {
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss");

        System.out.println("\n🎬 开始 Sora-2 视频生成完整流程");
        System.out.println("⏰ 开始时间: " + LocalDateTime.now().format(formatter));
        System.out.println();

        try {
            // 第一步：提交请求
            VideoSubmitResult result = step1SubmitRequest();

            // 第二步：轮询状态
            step2PollStatus(result.videoId);

            // 第三步：下载视频
            step3DownloadVideo(result.videoId, result.isImageMode);

            System.out.println("\n============================================================");
            System.out.println("🎉 完整流程执行成功!");
            System.out.println("============================================================");
            System.out.println("⏰ 结束时间: " + LocalDateTime.now().format(formatter));

        } catch (Exception e) {
            System.err.println("\n❌ 流程终止：" + e.getMessage());
            e.printStackTrace();
        }
    }

    // ========== 工具方法 ==========

    private static void addFormField(PrintWriter writer, String name, String value) {
        writer.append("--").append(BOUNDARY).append("\r\n");
        writer.append("Content-Disposition: form-data; name=\"").append(name).append("\"").append("\r\n");
        writer.append("\r\n");
        writer.append(value).append("\r\n");
    }

    private static void addFilePart(PrintWriter writer, OutputStream out, String fieldName, File file) throws IOException {
        String fileName = file.getName();
        writer.append("--").append(BOUNDARY).append("\r\n");
        writer.append("Content-Disposition: form-data; name=\"").append(fieldName)
              .append("\"; filename=\"").append(fileName).append("\"").append("\r\n");
        writer.append("Content-Type: ").append(Files.probeContentType(file.toPath())).append("\r\n");
        writer.append("\r\n");
        writer.flush();

        try (FileInputStream in = new FileInputStream(file)) {
            byte[] buffer = new byte[8192];
            int bytesRead;
            while ((bytesRead = in.read(buffer)) != -1) {
                out.write(buffer, 0, bytesRead);
            }
            out.flush();
        }

        writer.append("\r\n");
        writer.flush();
    }

    private static String readResponse(HttpURLConnection conn) throws IOException {
        try (BufferedReader reader = new BufferedReader(
                new InputStreamReader(
                    conn.getResponseCode() < 400 ? conn.getInputStream() : conn.getErrorStream()))) {
            StringBuilder response = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                response.append(line);
            }
            return response.toString();
        }
    }

    private static String formatTimestamp(long timestamp) {
        if (timestamp == 0) return "N/A";
        LocalDateTime dateTime = LocalDateTime.ofInstant(
            Instant.ofEpochSecond(timestamp), ZoneId.systemDefault());
        return dateTime.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"));
    }

    // 辅助类：提交结果
    private static class VideoSubmitResult {
        String videoId;
        boolean isImageMode;

        VideoSubmitResult(String videoId, boolean isImageMode) {
            this.videoId = videoId;
            this.isImageMode = isImageMode;
        }
    }
}
